### PR TITLE
Set 0 sync period in scheduler integration test

### PR DESCRIPTION
**Pull request by Huang-Wei**
**What type of PR is this?**

/kind cleanup

**What this PR does / why we need it**:

The resync period reconciles between the in-memory cache and the business logic. It's a misunderstanding that it's used to sync with apiserver. A common pattern is set it to 0 and widely used across the codebase.

This PR sets the resync period to 0.

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

/sig scheduling
/priority important-longterm

### CODE REVIEW

File: test/integration/scheduler/framework_test.go
Comment (keerthigd) [inline test/integration/scheduler/framework_test.go]: is it a good idea if you specify the sync period as `0` explicitly? rather than assuming the default value.

Comment (Huang-Wei) [reply test/integration/scheduler/framework_test.go]: Setting to `0` is a common pattern and I believe it's the default value. And again, the "sync period" is NOT syncing with apiserver. See:

- #75622
- https://github.com/kubernetes/kubernetes/issues/75495#issuecomment-475629555

Comment (Huang-Wei) [reply test/integration/scheduler/framework_test.go]: And setting to 0 doesn’t mean never resync, my memory is that it underneath sets it to a random 10ish minute.

**Review by hex108 — APPROVED** [review]
/lgtm

Thanks!

The detail for resync period could be found at https://groups.google.com/forum/#!msg/kubernetes-sig-api-machinery/PbSCXdLDno0/hEG1YykvDQAJ

**Comment by liggitt** [comment]
/hold

batches including this PR are flaking badly on scheduler integration tests:
* https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/batch/pull-kubernetes-integration/1171762472709787651/
* https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/batch/pull-kubernetes-integration/1171770526830432257/
* https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/batch/pull-kubernetes-integration/1171778580598755330/
* https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/batch/pull-kubernetes-integration/1171787137259933699/
* https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/batch/pull-kubernetes-integration/1171796195543093250/

please review closely to ensure this is not introducing a regression

**Comment by liggitt** [comment]
all scheduler-related PRs in the failing batches: https://github.com/kubernetes/kubernetes/pull/82187 https://github.com/kubernetes/kubernetes/pull/82209 https://github.com/kubernetes/kubernetes/pull/82222

**Comment by Huang-Wei** [comment]
This PR needs further work. I will take a close look.

**Comment by hex108** [comment]
/remove-lifecycle rotten

This patch should be OK. Maybe we could try it again?

**Comment by Huang-Wei** [comment]
> This patch should be OK. Maybe we could try it again?

Yes, the direction is what we need, just need to look into the test failures. I will find time to look into them.

**Comment by Huang-Wei** [comment]
Two tests [failed](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/82222/pull-kubernetes-integration/1321897136840773632) with the 1st commit: `TestBindPlugin` and `TestPreemptWithPermitPlugin`. I will use 2 comments explaining them separately.

**Comment by Huang-Wei** [comment]
The `TestBindPlugin` failure is actually a testing smell. The fake bind plugin fails the Binding phase on purpose and expects 4 events are trigger in a row: `fail bind (count=1)`, `fail reserve (count=1)`, `fail bind (count=2)`, `fail reserve (count=2)`. The latter two actually don't quite make sense - they're triggered by **rescheduling of the failed pod**. In the original test, it's assured by the forced resync (defaults to 1 second) of informers. So In other words, the failed Pod (rejected by bind plugin) will re-scheduled every resyncPeriod, and hence the bind plugin will also be called every resyncPeriod. This can be verified by passing the test by adding more expected events:

```go
{
	name:               "bind plugin fails to bind the pod",
	bindPluginStatuses: []*framework.Status{framework.NewStatus(framework.Error, "failed to bind"), framework.NewStatus(framework.Success, "")},
	expectInvokeEvents: []pluginInvokeEvent{
		{pluginName: bindPlugin1.Name(), val: 1},
		{pluginName: reservePlugin.Name(), val: 1},
		{pluginName: bindPlugin1.Name(), val: 2},
		{pluginName: reservePlugin.Name(), val: 2},
		{pluginName: bindPlugin1.Name(), val: 3},
		{pluginName: reservePlugin.Name(), val: 3},
		{pluginName: bindPlugin1.Name(), val: 4},
		{pluginName: reservePlugin.Name(), val: 4},
        ...
	},
},
```

I added a fix in the 2nd commit to remove checking for the unnecessary events.

**Comment by Huang-Wei** [comment]
The `TestPreemptWithPermitPlugin` failure is a bit tricky, let's read the error log:

```
13766 I1029 20:02:39.357896  121899 scheduling_queue.go:812] About to try and schedule pod preempt-with-permit-plugina79a70be-bce6-4368-80b4-8b8a5ff3fe4a/w      aiting-pod
13767 I1029 20:02:39.357908  121899 scheduler.go:438] Attempting to schedule pod: preempt-with-permit-plugina79a70be-bce6-4368-80b4-8b8a5ff3fe4a/waiting-po      d
13768 I1029 20:02:39.358046  121899 framework.go:876] one or more plugins asked to wait and no plugin rejected pod "waiting-pod"
13769 I1029 20:02:39.358087  121899 framework.go:899] pod "waiting-pod" waiting on permit
13770 I1029 20:02:39.371391  121899 eventhandlers.go:172] add event for unscheduled pod preempt-with-permit-plugina79a70be-bce6-4368-80b4-8b8a5ff3fe4a/pree      mptor-pod
13771 I1029 20:02:39.371439  121899 scheduling_queue.go:812] About to try and schedule pod preempt-with-permit-plugina79a70be-bce6-4368-80b4-8b8a5ff3fe4a/p      reemptor-pod
13772 I1029 20:02:39.371457  121899 scheduler.go:438] Attempting to schedule pod: preempt-with-permit-plugina79a70be-bce6-4368-80b4-8b8a5ff3fe4a/preemptor-      pod
13773 I1029 20:02:39.371640  121899 httplog.go:89] "HTTP" verb="POST" URI="/api/v1/namespaces/preempt-with-permit-plugina79a70be-bce6-4368-80b4-8b8a5ff3fe4      a/pods" latency="3.239556ms" userAgent="scheduler.test/v0.0.0 (linux/amd64) kubernetes/$Format" srcIP="127.0.0.1:36528" resp=201
13774 I1029 20:02:39.378701  121899 eventhandlers.go:204] delete event for unscheduled pod preempt-with-permit-plugina79a70be-bce6-4368-80b4-8b8a5ff3fe4a/w      aiting-pod
13775 I1029 20:02:39.378757  121899 framework.go:908] pod "waiting-pod" rejected while waiting on permit: removed
13776 E1029 20:02:39.378840  121899 factory.go:345] "Error scheduling pod; retrying" err="pod \"waiting-pod\" rejected while waiting on permit: removed" po      d="preempt-with-permit-plugina79a70be-bce6-4368-80b4-8b8a5ff3fe4a/waiting-pod"
13777 W1029 20:02:39.378856  121899 factory.go:351] Pod preempt-with-permit-plugina79a70be-bce6-4368-80b4-8b8a5ff3fe4a/waiting-pod doesn't exist in informe      r cache: pod "waiting-pod" not found
13778 I1029 20:02:39.378882  121899 scheduler.go:334] Updating pod condition for preempt-with-permit-plugina79a70be-bce6-4368-80b4-8b8a5ff3fe4a/waiting-pod       to (PodScheduled==False, Reason=Unschedulable)
13779 I1029 20:02:39.379006  121899 httplog.go:89] "HTTP" verb="DELETE" URI="/api/v1/namespaces/preempt-with-permit-plugina79a70be-bce6-4368-80b4-8b8a5ff3f      e4a/pods/waiting-pod" latency="6.850599ms" userAgent="scheduler.test/v0.0.0 (linux/amd64) kubernetes/$Format" srcIP="127.0.0.1:36528" resp=200
13780 I1029 20:02:39.379278  121899 factory.go:329] "Unable to schedule pod; no fit; waiting" pod="preempt-with-permit-plugina79a70be-bce6-4368-80b4-8b8a5f      f3fe4a/preemptor-pod" err="0/1 nodes are available: 1 Insufficient cpu, 1 Insufficient memory."
13781 I1029 20:02:39.379373  121899 scheduler.go:334] Updating pod condition for preempt-with-permit-plugina79a70be-bce6-4368-80b4-8b8a5ff3fe4a/preemptor-p      od to (PodScheduled==False, Reason=Unschedulable)
```

It shows that the waiting pod is expected to be assumed and hence occupy a certain amount of resource, and also passed the Permit() phase to wait there. Afterwards, the preemptor pod triggers a preemption. However, note that the waiting pod is an **unassigned Pod** (without .spec.nodeName set), so it triggers `sched.deletePodFromSchedulingQueue()` instead of `sched.deletePodFromCache()`. The minor difference results in the preemptor pod has no chance to be moved back to activeQ/backoffQ - this can be verified by searching the entry of "Attempting to schedule pod ... preemptor-pod".

Well, why the original test can pass? It's because the force resync acted as the factor to trigger moving preemptor to be rescheduled.

In the 2nd commit, I commented out the check to verify the preemptor to be scheduled. This is a temporary fix, in an ideal case, we may come up with a generic and smart way to trigger `MoveAllToActiveOrBackoffQueue()` - e.g., upon deactivating Pod from assumed pods cache, etc.

**Comment by Huang-Wei** [comment]
cc/ @ahg-g @alculquicondor @adtac @hex108

**Comment by Huang-Wei** [comment]
Successful integration test runnings: [1](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/82222/pull-kubernetes-integration/1321967191058288640)

**Comment by alculquicondor** [comment]
> However, note that the waiting pod is an **unassigned Pod** (without .spec.nodeName set), so it triggers `sched.deletePodFromSchedulingQueue()` instead of `sched.deletePodFromCache()`

A bit of a tangent, but how does this clear the assumed Pod from the cache? I don't see any calls to the cache from `deletePodFromSchedulingQueue`.

**Review by alculquicondor — COMMENTED** [review]
/approve

File: test/integration/scheduler/framework_test.go
Comment (adtac) [inline test/integration/scheduler/framework_test.go:1888]:
```suggestion
	if err := wait.Poll(200*time.Millisecond, wait.ForeverTestTimeout, func() (bool, error) {
```

File: test/integration/scheduler/framework_test.go
Comment (hex108) [reply test/integration/scheduler/framework_test.go:1221]: `MoveAllToActiveOrBackoffQueue` will be called when forceing sync, so it will be triggered?

Comment (Huang-Wei) [reply test/integration/scheduler/framework_test.go:1221]: I think it triggered the following logic in `sched.updatePodInSchedulingQueue()` to reschedule the failed pod:

https://github.com/kubernetes/kubernetes/blob/24a9d07bf810c846a79444cb4ea4adee90d4a821/pkg/scheduler/internal/queue/scheduling_queue.go#L447-L450

Comment (hex108) [reply test/integration/scheduler/framework_test.go:1221]: @Huang-Wei Yes, you're right. :)

**Comment by Huang-Wei** [comment]
Still holding this PR, I will resolve comments after #96062 and #96071 get merged.

**Comment by liggitt** [comment]
/hold

is this making the integration tests run in a way that does not match the real kube-scheduler environment? The analysis in https://github.com/kubernetes/kubernetes/pull/96071#issuecomment-720955648 seems to indicate the event handlers the scheduler is attaching to the informer do not handle update events correctly. Is this PR masking that bug?

**Comment by Huang-Wei** [comment]
> is this making the integration tests run in a way that does not match the real kube-scheduler environment?

Nope, kube-scheduler always adopts the 0 resync period for a while:

https://github.com/kubernetes/kubernetes/blob/a439bc55724f6560c14bf4e025946c0e6312629b/cmd/kube-scheduler/app/options/options.go#L289

It's just the integration tests that were written in a "1-second-resync" style.

**Comment by liggitt** [comment]
ok, thanks

/hold cancel

File: test/integration/scheduler/framework_test.go
Comment (Huang-Wei) [reply test/integration/scheduler/framework_test.go]: Done.

File: test/integration/scheduler/framework_test.go
Comment (Huang-Wei) [reply test/integration/scheduler/framework_test.go]: Done.

**Comment by Huang-Wei** [comment]
Hmm.. it seems this PR falls into a weird CI bug.

**Comment by Huang-Wei** [comment]
@alculquicondor @adtac comments have been addressed. PTAL.

**Comment by Huang-Wei** [comment]
@liggitt would you mind /approve the changes in `test/integration/util/util.go`? In terms of scheduler, we don't need a non-zero syncPeriod option in integrate tests.

**Comment by Huang-Wei** [comment]
Kindly ping @liggitt for /approve.

**Pull request merged**
